File: paperparcel-java7/src/main/java/nz/bradcampbell/paperparcel/PaperParcelable.java
package nz.bradcampbell.paperparcel;

import android.os.Parcel;
import android.os.Parcelable;

public abstract class PaperParcelable implements Parcelable {
  public static class Creator<T> implements Parcelable.Creator<T> {
    private final Class<? extends T> type;

    public Creator(Class<? extends T> type) {
      this.type = type;
    }

    public T createFromParcel(Parcel parcel) {
      return PaperParcels.unwrap(parcel.readParcelable(type.getClassLoader()));
    }

    @Override public T[] newArray(int i) {
      return PaperParcels.newArray(type, i);
    }
  }

  @Override public int describeContents() {
    return 0;
  }

  @Override public void writeToParcel(Parcel parcel, int flags) {
    parcel.writeParcelable(PaperParcels.wrap(this), flags);
  }
}

File: paperparcel-kotlin/src/main/kotlin/nz/bradcampbell/paperparcel/PaperParcelable.kt
package nz.bradcampbell.paperparcel

import android.os.Parcel
import android.os.Parcelable

interface PaperParcelable : Parcelable {
  class Creator<T>(private val type: Class<out T>) : Parcelable.Creator<T> {
    override fun createFromParcel(parcel: Parcel): T {
      return PaperParcels.unwrap<T>(parcel.readParcelable<Parcelable>(type.classLoader))
    }

    override fun newArray(i: Int): Array<T> {
      return PaperParcels.newArray(type, i)
    }
  }

  override fun writeToParcel(outParcel: Parcel, flags: Int) {
    outParcel.writeParcelable(PaperParcels.wrap(this), flags)
  }

  override fun describeContents(): Int {
    return 0
  }
}
